pkg/log/log.go

```diff
@@ -0,0 +1,89 @@
+package log
+
+import (
+	"flag"
+	"fmt"
+	"io"
+
+	"github.com/operator-framework/operator-sdk/pkg/log/zap"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"github.com/gin-gonic/gin"
+	"github.com/go-logr/logr"
+	"github.com/spf13/pflag"
+)
+
+var (
+	log logr.Logger
+)
+
+// InitializeLogger initializes the logger.
+func InitializeLogger(withName string) {
+	log = logf.Log.WithName(withName)
+
+	zapFlagSet := pflag.NewFlagSet("zap", pflag.ExitOnError)
+
+	// Add the zap logger flag set to the CLI. The flag set must
+	// be added before calling pflag.Parse().
+	pflag.CommandLine.AddFlagSet(zapFlagSet)
+
+	// Add flags registered by imported packages (e.g. glog and
+	// controller-runtime)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+
+	pflag.Parse()
+
+	// Use a zap logr.Logger implementation. If none of the zap
+	// flags are configured (or if the zap flag set is not being
+	// used), this defaults to a production zap logger.
+	//
+	// The logger instantiated here can be changed to any logger
+	// implementing the logr.Logger interface. This logger will
+	// be propagated through the whole operator, generating
+	// uniform and structured logs.
+	logf.SetLogger(zap.Logger())
+}
+
+func SetOutput(out io.Writer, isTestingMode bool, withName string) {
```

Comment:
May I know what is the purpose of this function?

Reply:
it's to be able to redirect our outpit if we don't want it to go to os.Stderr. This would be useful for testing i guess. It also allows me to write tests for these functions by redirecting the output to a bytes.Buffer..

Reply:
So, then we don't need an additional method in the prod logger just for the sake of using it in tests. I would rather initialize a test logger with the needed redirection/setup at the beginning of the test(s). I believe that you don't need to do this dance every time:
```go
logger.SetOutput(&buf, true, "logger_tests")
defer func() {
	logger.SetOutput(os.Stderr, false, "logger_tests")
}()
```
just create an interface for the current logger and create two implementations - the current one and one for tests.

Reply:
Actually i didn't need to do it in each sub-test - i moved out so this is fixed here: a873ba650aef35eec3abafe2d036c905c8aeec13